### PR TITLE
Transition from manifest v2 to v3 for chrome

### PR DESCRIPTION
**Pull request by TarikGul**
closes: https://github.com/polkadot-js/extension/issues/1316
closes: https://github.com/polkadot-js/extension/issues/310

## Summary (**BREAKING CHANGES**)

The following PR gives support for manifest V3. Below will help document the exact changes that have taken place in this PR and how it affects the interface given to users.

## `@polkadot/extension-base`

The following covers exported items from `@polkadot/extension-base`. 

### `class Extension`

- `Extension.accountsForget`: This is now asynchronous, and has a return type of `Promise<boolean>` instead of `boolean`.
- `Extension.authorizeUpdate`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.
- `Extension.metadataApprove`: This is now asynchronous, and has a return type of `Promise<boolean>` instead of `boolean`.
- `Extension.removeAuthorization`: This is now asynchronous, and has a return type of `Promise<ResponseAuthorizeList>` instead of `ResponseAuthorizeList`.

### `class State`

- Initializing `new State(providers)` has now changed. This is because `extractMetadata` within the constructor has to be asynchronous now. So we instead have the logic from the constructor and placed it within a init function to initialize state. Example below:

```typescript
// Old
const state = new State(providers);

// New
const state = new State(providers);
await state.init();
```

- `State.saveCurrentAuthList`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.
- `State.saveDefaultAuthAccounts`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.
- `State.updateDefaultAuthAccounts`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.
- `State.removeAuthorization`: This is now asynchronous, and has a return type of `Promise<AuthUrls>` instead of `AuthUrls`.
- `State.updateAuthorizedAccounts`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.
- `State.saveMetadata`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.

### `class AccountsStore`

- `AccountsStore.set`: This is now asynchronous, and has a return type of `Promise<void>` instead of `void`.

### `class BaseStore`

- `BaseStore.all`: This is now asynchronous, and has a return type of `Promise<void> instead of `void`.
- `BaseStore.allMap`: This is now asynchronous, and has a return type of `Promise<void> instead of `void`.
- `BaseStore.get`: This is now asynchronous, and has a return type of `Promise<void> instead of `void`.
- `BaseStore.remove`: This is now asynchronous, and has a return type of `Promise<void> instead of `void`.
- `BaseStore.set`: This is now asynchronous, and has a return type of `Promise<void> instead of `void`.

### SIgnArea Component

Now uses `alarms`

### ToastProvider component

Now uses `alarms`

## [Checklist](https://developer.chrome.com/docs/extensions/develop/migrate/checklist)

- [X] [Update the manifest](https://developer.chrome.com/docs/extensions/develop/migrate/checklist#update_the_manifest)
    - [X] [Change the manifest version from 2 to 3](https://developer.chrome.com/docs/extensions/develop/migrate/manifest#change-version)
    - [X] [Update Host Permissions](https://developer.chrome.com/docs/extensions/develop/migrate/manifest#update-host-permissions)
    - [X] [Update `web_accessible_resources`](https://developer.chrome.com/docs/extensions/develop/migrate/manifest#update-wa-resources)
- [x] [Migrate to service worker](https://developer.chrome.com/docs/extensions/develop/migrate/checklist#migrate_to_a_service_worker)
    - [x] [Update the "background" field in the manifest](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#update-bg-field)
    - [x] [Move DOM and window calls to an offscreen document](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#move-dom-and-window)
        - This one is a bit annoying since it requires us to use the offscreen API which requires the `offscreen` permission. See [update_permissions](https://developer.chrome.com/docs/extensions/develop/concepts/permission-warnings#update_permissions), and [keep-features](https://developer.chrome.com/docs/extensions/develop/migrate#keep-features) for more info. 
        I believe this is actually not necessary given the specific constraints.
    - [x] [Convert localStorage to chrome.storage.local](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#convert-localstorage)
    - [x] [Register listeners synchronously](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#register-listeners)
        - We don't have any synchronous listeners that use any browser actions so I think this is okay to check off.
    - [x] [Replace calls to XMLHttpRequest() with global fetch()](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#replace-xmlhttprequest).
        - We don't use `XMLHttpRequest`
    - [x] [Persist states](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#persist-states)
    - [x] [Convert timers to alarms](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#convert-timers)
        - There are a few places where we use setTimeout, not sure if that is applicable here and we need to switch it to `await chrome.alarms.create`
    - [x] [Keep the service worker alive (in exceptional cases)](https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#keep-sw-alive) 
- [x] [Update API calls](https://developer.chrome.com/docs/extensions/develop/migrate/checklist#update_api_calls)
    - [x] [Replace tabs.executeScript() with scripting.executeScript()](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-executescript).
        - We don't use `tabs.executeScript()`
    - [x] [Replace tabs.insertCSS() and tabs.removeCSS() with scripting.insertCSS() and scripting.removeCSS()](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-insertcss-removecss).
        - We don't use `tabs.insertCSS()` or `tabs.removeCSS()`
    - [x] [Replace Browser Actions and Page Actions with Actions](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-browser-page-actions)
    - [x] [Replace functions that expect a Manifest V2 background context](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-mv2-function).
        - We don't use calls that expect a background context.
    - [x] [Replace callbacks with promises](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-callbacks)
    - [x] [Replace unsupported APIs](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-unsupported-apis)
- [x] [Replace web request listeners](https://developer.chrome.com/docs/extensions/develop/migrate/checklist#replace_blocking_web_request_listeners)
    - [x] [Update permissions](https://developer.chrome.com/docs/extensions/develop/migrate/blocking-web-requests#update-permissions)
    - [x] [Create declarative net request rules](https://developer.chrome.com/docs/extensions/develop/migrate/blocking-web-requests#create-dnr-rules)
    - We don't use declarative net requests
    After completing the items above, you may want to review a few [common use cases](https://developer.chrome.com/docs/extensions/develop/migrate/blocking-web-requests#common-use-cases):
    - [x] [Block a single URL](https://developer.chrome.com/docs/extensions/develop/migrate/blocking-web-requests#block-a-single-url)
    - [x] [Redirect-multiple-urls](https://developer.chrome.com/docs/extensions/develop/migrate/blocking-web-requests#redirect-multiple-urls)
    - [x] [Block cookies](https://developer.chrome.com/docs/extensions/develop/migrate/blocking-web-requests#block-cookies)
- [x] [Improve extension security](https://developer.chrome.com/docs/extensions/develop/migrate/checklist#improve_extension_security)
    - [x] [Remove execution of arbitrary strings](https://developer.chrome.com/docs/extensions/develop/migrate/improve-security#remove-execution-of-strings).
    - [x] [Remove remotely hosted code](https://developer.chrome.com/docs/extensions/develop/migrate/improve-security#remove-remote-code)
    - [x] [Update content security policy](https://developer.chrome.com/docs/extensions/develop/migrate/improve-security#update-csp).
    - [x] [Remove unsupported content security policy values](https://developer.chrome.com/docs/extensions/develop/migrate/improve-security#remove-unsupported-csv)

### CODE REVIEW

**Comment by TarikGul** [comment]
So far the extension is working fine in chrome with the exception of the following errors: 

```
Error: Invalid url chrome://extensions/?errors=fnpncejekijbiiemffncneiiilejbmbj, expected to start with http: or https: or ipfs: or ipns:
```

**Comment by TarikGul** [comment]
Instead of causing breaking changes to `extension-base` by enforcing async at the top level of State, I am going to just use the callback method for `storage.local`

**Review by Jrboss999 — APPROVED** [review]

**Comment by TarikGul** [comment]
@Tbaut Yea the I was able to fix the localStorage error but then reverted my changes because I would like to change the design of the implementation, but yea the localStorage issue will be fixed soon :)

**Comment by adamdossa** [comment]
Is it expected that this PR will be merged in the near future? We have a downstream dependency for our Polymesh Wallet which requires this in order to move to manifest V3. @Tbaut

**Comment by TarikGul** [comment]
> Is it expected that this PR will be merged in the near future? We have a downstream dependency for our Polymesh Wallet which requires this in order to move to manifest V3. @Tbaut

Yes this is absolutely a priority! My goal is to get it done this week.

**Comment by Tbaut** [comment]
Hello there, I tried this branch and documented my tests:

- loaded extension with manifest v2 (master v0.48.2) imported Bob and Dave
- did a tx with Bob
- exported Bob
- removed Bob
- checkout tv-manifest-v3, `yarn && yarn build` refresh the extension in chrome
  BTW just to be 100% sure this is indeed loading the extension I wantwith v3, I bumped the version in `packages/extension/package.json` before building, and made sure the new version is displayed in chrome.
- all my accounts are here as expected
- I tested successfully:
  - import an account with mnemonic and derivation
  - import an account with JSON
  - export a batch of accounts
  - import a batch of accounts
  - forget an account
  - sign a TX
  - update the metadata
  - imported an account with Polkadot Vault
  - signed a tx with Polkadot Vault
  - forget an account with Polkadot Vault

Note that this includes connecting to some websites, and changing the connected accounts.

Somewhat unexpected for normal users
- on a website where I had already allowed Bob with v2, I had to authenticate again.

In summary, it's looking amazing :clap:

File: packages/extension-ui/src/Popup/Signing/Request/SignArea.tsx [hunk 1]
```diff
@@ -30,25 +32,41 @@ function SignArea ({ buttonText, className, error, isExternal, isFirst, setError
...
-        !isLocked && setSavePass(true);
-      })
-      .catch((error: Error) => console.error(error));
-
-    return () => {
-      !!timeout && clearTimeout(timeout);
+    const lockSigner = async () => {
+      setIsLocked(null);
+
+      const { isLocked, remainingTime } = await isSignLocked(signId);
```
[the remark this answers is not in the document]
Comment (TarikGul) [reply packages/extension-ui/src/Popup/Signing/Request/SignArea.tsx:38]: Also getting this error: 

```
Error clearing the alarm: TypeError: Cannot read properties of undefined (reading 'create')
```

**Review by bee344 — APPROVED** [review]
Great job
